### ISSUE TITLE
Why does IsStructurallyValidLanguageTag allow duplicate extension key values?

### ISSUE THREAD

**Issue by gsathya** [tc39/ecma402#262]
What's the main use case for this? It seems to me that it'd be a better user experience to provide an error earlier instead?

It'd make implementations easier. We can just call out to ICU (which has the behavior of throwing) and not do all the complicated string manipulations in [UnicodeExtensionValue](https://tc39.github.io/ecma402/#sec-unicodeextensionvalue) helping performance/memory.

Chakra and V8 throw, whereas JSC and Spidermonkey don't.

```
➜  eshost -se 'new Intl.Collator("de-u-co-phonebk-co-emoji").resolvedOptions().locale'      
     
#### JavaScriptCore, SpiderMonkey        
de-u-co-phonebk

#### V8                                                                                                 
RangeError: Invalid language tag: de-u-co-phonebk-co-emoji                                                             

#### Chakra                                                                                                                                             
RangeError: Locale 'de-u-co-phonebk-co-emoji' is not well-formed           
```

**Comment by littledan** [comment]
Good catch! Maybe we should discuss this at the Intl meeting.

**Comment by jungshik** [comment]
The reason for allowing duplicate sub-tag (key) in the Unicode extension and ignoring the 2nd or later key-value is, I guess, the following in [RFC 6067](https://tools.ietf.org/html/rfc6067) defining 'U' extension. 

```
  Only the first occurrence of an attribute or key conveys meaning in a
   language tag.  When interpreting tags containing the Unicode locale
   extension, duplicate attributes or keywords are ignored in the
   following way: ignore any attribute that has already appeared in the
   tag and ignore any keyword whose key has already occurred in the tag.
```

Nonetheless, I'm inclined to agree with @gsathya . Treating a language tag with duplicate keyword in U extension would simplify the implementation. ICU's uloc_forLanguageTag yields 'illegal argument' error for a language tag with a duplicate keyword.

**Comment by jungshik** [comment]
Another reason for allowing duplicate keywords  comes straight from [BCP 47]{https://tools.ietf.org/html/bcp47}.  BCP47 has the following for duplicate variant subtags and duplicate singleton extension subtags.

```
There are no duplicate variant subtags.
There are no duplicate singleton (extension) subtags.
```

But, for subtags in extension, it has the following: 

```
A tag is considered valid for a given extension (Section 3.7) (as of
a particular version, revision, and date) if it meets the criteria
for "valid" above and also satisfies this condition:

 Each subtag used in the extension part of the tag is valid
 according to the extension.
```

Nonetheless, my inclination remains the same. A minor worry is an incompatibility between Ecma 402 locale validation and other implementation.

**Comment by anba** [comment]
I don't understand why this issue wasn't filed as a bug against ICU4C at <https://unicode-org.atlassian.net/secure/MyJiraHome.jspa>? For example ICU4J allows duplicate keywords:

```java
public class DuplicateUnicodeKeyword {
    public static void main(String[] args) {
        var loc = com.ibm.icu.util.ULocale.forLanguageTag("en-u-kf-true-kf-false");
        System.out.println(loc.toLanguageTag()); // Prints "en-u-kf-true"
    }
}
```

Source: https://github.com/unicode-org/icu/blob/1a6f0bf7681ed84efce8f8c06b9cfb1ded983683/icu4j/main/classes/core/src/com/ibm/icu/impl/locale/InternalLocaleBuilder.java#L555

**Comment by jungshik** [comment]
@anba, didn't you read my comment about  RFC 6067 ? It does allows duplicate keywords.

**Comment by jungshik** [comment]
@anba, sorry for my stupidity. I misunderstood your question. Filing a bug against ICU makes sense instead of departing from RFC 6067.  I'll do that.

**Comment by jungshik** [comment]
Filed https://unicode-org.atlassian.net/browse/ICU-20140 .

**Comment by jungshik** [comment]
The above ICU issue was  fixed by unicode-org/icu#136. 

This issue can be closed now.

**Comment by littledan** [comment]
Closing per https://github.com/tc39/ecma402/issues/262#issuecomment-423040235 and confirmed in the September 2018 Intl call.
